### PR TITLE
Complete re-write of "Transitive persistence / Cascade Operations"

### PR DESCRIPTION
**Pull request by ThomasLandauer**
Follow-up of https://github.com/doctrine/doctrine2/pull/6171

### CODE REVIEW

File: docs/en/reference/working-with-associations.rst [hunk 1]
```diff
@@ -488,30 +471,52 @@ and the "remove" operation.
...
+...you can now create a user and an associated comment like this:
 
 .. code-block:: php
 
     <?php
     $user = new User();
-    $myFirstComment = new Comment();
-    $user->addComment($myFirstComment);
-    $myFirstComment->setUser($user);
+    $user->comment('Lorem ipsum', new DateTime());
```
Comment (alcaeus) [inline docs/en/reference/working-with-associations.rst]: The example you're replacing sets the `user` property in the comment, but the `comment` method written above does not do that. You should add that to the example.

File: docs/en/reference/working-with-associations.rst [hunk 2]
```diff
@@ -488,30 +471,52 @@ and the "remove" operation.
...
     
     $em->persist($user);
     $em->flush();
 
+Thanks to ``cascade: remove``, you can easily delete a user too (without having to loop through all associated comments):
```
Comment (alcaeus) [inline docs/en/reference/working-with-associations.rst]: I'd reword this:
> Using ``cascade: remove`` you can delete a user and all linked comments without looping through them:

File: docs/en/reference/working-with-associations.rst [hunk 3]
```diff
@@ -488,30 +471,52 @@ and the "remove" operation.
...
+    are fetched into memory (even if they are marked as lazy) when
+    the cascade operation is about to be performed. However this approach allows
+    entity lifecycle events to be performed for each of these operations.
+
+    However, pulling objects graph into memory on cascade can cause considerable performance
+    overhead, especially when the cascaded collections are large. Make sure
+    to weigh the benefits and downsides of each cascade operation that you define.
+
+    To rely on the database level cascade operations for the delete operation instead, you can
+    configure each join column with the **onDelete** option. See the respective
```
Comment (mikeSimonson) [inline docs/en/reference/working-with-associations.rst]: I would add the link to the section here.

File: docs/en/reference/working-with-associations.rst [hunk 4]
```diff
@@ -488,30 +471,52 @@ and the "remove" operation.
...
+
+    $em->remove($user);
+    $em->flush();
+
 .. note::
 
-    This code does not associate the newly created comment with the user.
-    To achieve this, you *always* have to call ``$myFirstComment->setUser($user);`` –
-    no matter if ``cascade: persist`` is set or not.
+    The idea of ``cascade: persist`` is not to save you any lines of code in the controller.
```
Comment (alcaeus) [inline docs/en/reference/working-with-associations.rst]: I'd move the note about `cascade: persist` up above the section on `cascade: remove` to keep the topics more separated.

**Review by alcaeus — CHANGES_REQUESTED** [review]
Couple of minor errors. Since you're reworking this section of the documentation, could you also document the other cascade actions? `merge`, `detach` and `refresh` are not explained at all in this chapter.

File: docs/en/reference/working-with-associations.rst [hunk 5]
```diff
@@ -444,34 +427,34 @@ following code:
...
+            $this->id = User::new();
+            $this->comments = new ArrayCollection();
+        }
 
-To have Doctrine handle both cases automatically we can change the
-``User#commentsAuthored`` property to cascade both the "persist"
-and the "remove" operation.
+        public function comment(string $text, DateTimeInterface $time) : void
+        {
+            $this->comments->add(Comment::create($text, $time));
```
Comment (mikeSimonson) [inline docs/en/reference/working-with-associations.rst]: I would add the $comment->setUser() here otherwise I can already see them copy pasting this.

File: docs/en/reference/working-with-associations.rst (hunk 3 above)
Comment (ThomasLandauer) [inline docs/en/reference/working-with-associations.rst]: I just copied this box 1:1 from before. If you tell me the link, I'll certainly include it ;-)

File: docs/en/reference/working-with-associations.rst (hunk 1 above)
Comment (ThomasLandauer) [inline docs/en/reference/working-with-associations.rst]: I modified the `comment()` method. Not sure, if this is the correct way, though. Please double-check!

File: docs/en/reference/working-with-associations.rst (hunk 2 above)
Comment (ThomasLandauer) [inline docs/en/reference/working-with-associations.rst]: OK, done.  
I kept the "Thanks" part, to indicate that the remove-cascading has in fact alread been set in the example above.

File: docs/en/reference/working-with-associations.rst (hunk 4 above)
Comment (ThomasLandauer) [inline docs/en/reference/working-with-associations.rst]: Done :-)

File: docs/en/reference/working-with-associations.rst (hunk 5 above)
Comment (ThomasLandauer) [inline docs/en/reference/working-with-associations.rst]: Done, but please double-check!

**Comment by ThomasLandauer** [comment]
@alcaeus About `merge`, `detach` and `refresh`: No, sorry :-(  
I don't have any idea how they work. I'm glad I finally understood `persist` well enough to explain it (thanks to the patience of @Ocramius )

File: docs/en/reference/working-with-associations.rst (hunk 1 above)
Comment (alcaeus) [inline docs/en/reference/working-with-associations.rst]: Looks good!

File: docs/en/reference/working-with-associations.rst (hunk 3 above)
Comment (mikeSimonson) [inline docs/en/reference/working-with-associations.rst]: :doc:`the **onDelete** option <working-with-objects>`

Comment (ThomasLandauer) [inline docs/en/reference/working-with-associations.rst]: Thanks! Done :-)
